sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanBuckets.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.collect.EvictingQueue;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.Status.CanonicalCode;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.Map;
+
+final class TracezSpanBuckets {
+  private static final int NUM_SAMPLES_PER_LATENCY_BUCKET = 10;
+  private static final int NUM_SAMPLES_PER_ERROR_BUCKET = 5;
+
+  private final Map<LatencyBoundaries, EvictingQueue<ReadableSpan>> latencyBuckets =
+      new HashMap<>();
+  private final Map<CanonicalCode, EvictingQueue<ReadableSpan>> errorBuckets = new HashMap<>();
+
+  public TracezSpanBuckets() {
+    for (LatencyBoundaries bucket : LatencyBoundaries.values()) {
+      latencyBuckets.put(
+          bucket, EvictingQueue.<ReadableSpan>create(NUM_SAMPLES_PER_LATENCY_BUCKET));
+    }
+    for (CanonicalCode code : CanonicalCode.values()) {
+      if (!code.toStatus().isOk()) {
+        errorBuckets.put(code, EvictingQueue.<ReadableSpan>create(NUM_SAMPLES_PER_ERROR_BUCKET));
+      }
+    }
+  }
+
+  public void addToBucket(ReadableSpan span) {
+    Status status = span.toSpanData().getStatus();
+    if (status.isOk()) {
+      long latency = span.getLatencyNanos();
+      for (LatencyBoundaries bucket : LatencyBoundaries.values()) {
+        if (latency >= bucket.getLatencyLowerBound() && latency < bucket.getLatencyUpperBound()) {
+          synchronized (latencyBuckets) {
+            latencyBuckets.get(bucket).add(span);
```

Comment:
Though I am worried about the performance impact of this synchronization, if it's just one latency bucket it's better, but I think it's pretty common for a server to only have a single method and serve huge QPS with similar latency (think an authentication server). The contention here seems like it could get pretty huge, possibly breaking the tenet that observability must not be intrusive.

It seems like a pretty small amount of code to implement a non-blocking ring buffer, if not in this PR can you make sure to try it in another that also adds a JMH? I don't think there's a need to have the bucket size match opencensus if there's a reason we find for a different.

```java
final class SpanBucket {
  // power of 2 means Integer.MAX_VALUE % BUCKET_SIZE = BUCKET_SIZE - 1 the maximum remainder, and
  // we will loop back to 0.
  private static final int BUCKET_SIZE = 16; 
  private final Span[] spans = new Span[BUCKET_SIZE];
  private final AtomicInteger index = new AtomicInteger();

  void add(Span span) {
    spans[UnsignedInts.remainder(index.getAndIncrement(), BUCKET_SIZE)] = span;
  }

  int size() {
    for (int i = BUCKET_SIZE - 1; i >= 0; i--) {
      if (spans[i] != null) {
        return i + 1;
      }
    }
    return 0;
  }
  
  void addTo(List<Span> result) {
    for (int i = 0; i < BUCKET_SIZE; i++) {
      Span span = spans[i];
      if (span != null) {
        result.add(span);
      } else {
        break;
      }
    }
  }
```

Reply:
Added a `SpanBucket` class to replace the `EvictingQueue`!